rfd/0abc-kube-access-forward-identity.md

```diff
@@ -0,0 +1,181 @@
+using X.509 certificate. Kubernetes Proxy will then forward the request to the
+Teleport service responsible for serving the Kubernetes cluster (Kube Service) or
+to the remote Teleport proxy when the Kubernetes Cluster belongs to a trusted cluster.
+For the proxy to be able to forward the user's identity to the correct service,
+it needs to generate a new X.509 key-pair with the user identity embedded that
+will be used during the mTLS authentication.
+The upstream service will then use the user identity to authorize the request and to
+impersonate the configured Kubernetes principals when forwarding the request to
+the Kubernetes API server.
+
+To avoid having to generate a new certificate for each request, the proxy will
+generate a new certificate for each user's identity and cache it in memory for
+`min(3h, certExpiration)`. This certificate cache must be keyed with a key that
+uniquely identifies the user's identity so that the proxy can retrieve the certificate
+for the correct identity when forwarding the request. In the past, we had situations
+where the cache was not properly keyed and the proxy would forward the request
+with the wrong certificate ending up with authorization errors.
+
+Currently, the key used to cache the certificate is the user's identity has the
+following format:
+
+```go
+func (c *authContext) key() string {
+	// it is important that the context key contains user, kubernetes groups and certificate expiry,
+	// so that new logins with different parameters will not reuse this context
+	return fmt.Sprintf("%v:%v:%v:%v:%v:%v:%v", c.teleportCluster.name, c.User.GetName(), c.kubeUsers, c.kubeGroups, c.kubeClusterName, c.certExpires.Unix(), c.Identity.GetIdentity().ActiveRequests)
+}
+```
+
+Each time we introduce a new field to the user's identity, we need to make sure
+that the key is updated to include the new field if the field is used by Kubernetes
+Access. This process is error-prone and it's easy to forget to update the key, and
+eventually introduce a bug or a security vulnerability.
+
+This RFD proposes a way to forward the user's identity from the proxy to the
+Kube Service and remote Teleport proxy that does not require the proxy to generate
+a new certificate for each user's identity.
+
+## Details
+
+The Proxy will authenticate to the Kube Service and remote Teleport proxy using its
+certificate. The proxy certificate does not contain any user identification
+and it's signed by the Teleport CA. The proxy will then forward the user's identity
+using HTTP headers or Proxy Protocol extensions.
+
+This change will remove the requirement for the proxy to generate a new certificate
+and thus call auth's `ProcessKubeCSR` endpoint to sign the key-pair containing the
+user's identity.
+
+To keep a chain of trust between the proxy and the upstream service, the
+user's identity forwarded by the proxy will be signed by the Teleport Proxy CA.
```

Comment:
What is Teleport proxy CA? Are we going to introduce a new CA?

Reply:
No, it was a typo. I mean the proxy private key and wrote CA